src/three-components/SmoothControls.ts

```diff
@@ -0,0 +1,533 @@
+
+  private[$velocity]: number = 0;
+
+  readonly sceneOrigin: Vector3 = new Vector3();
+
+  constructor(readonly camera: Camera, readonly element: HTMLElement) {
+    super();
+
+    this[$upQuaternion].setFromUnitVectors(camera.up, UP);
+    this[$upQuaternionInverse].copy(this[$upQuaternion]).inverse();
+
+    this[$onMouseMove] = (event: Event) =>
+        this[$handlePointerMove](event as MouseEvent);
+    this[$onMouseDown] = (event: Event) =>
+        this[$handlePointerDown](event as MouseEvent);
+    this[$onMouseUp] = (event: Event) =>
+        this[$handlePointerUp](event as MouseEvent);
+    this[$onWheel] = (event: Event) => this[$handleWheel](event as WheelEvent);
+    this[$onKeyDown] = (event: Event) =>
+        this[$handleKey](event as KeyboardEvent);
+    this[$onTouchStart] = (event: Event) =>
+        this[$handlePointerDown](event as TouchEvent);
+    this[$onTouchEnd] = (event: Event) =>
+        this[$handlePointerUp](event as TouchEvent);
+    this[$onTouchMove] = (event: Event) =>
+        this[$handlePointerMove](event as TouchEvent);
+
+    element.addEventListener('mousemove', this[$onMouseMove]);
+    element.addEventListener('mousedown', this[$onMouseDown]);
+    element.addEventListener('wheel', this[$onWheel]);
+    element.addEventListener('keydown', this[$onKeyDown]);
+    element.addEventListener('touchstart', this[$onTouchStart]);
+    element.addEventListener('touchmove', this[$onTouchMove]);
+
+    self.addEventListener('mouseup', this[$onMouseUp]);
+    self.addEventListener('touchend', this[$onTouchEnd]);
+
+    element.style.cursor = 'grab';
+
+    this[$previousPosition].copy(this.camera.position);
+    this[$positionToSpherical](this.camera.position, this[$spherical]);
+    this[$targetSpherical].copy(this[$spherical]);
+
+    this[$constraints] = Object.assign({}, DEFAULT_CONSTRAINTS);
+
+    this.setOrbit();
+  }
+
+  /**
+   * Clean up event handlers that are added to the configured element. Invoke
+   * this when getting rid of the controls, otherwise listeners will leak!
+   */
+  dispose() {
+    this.element.removeEventListener('mousemove', this[$onMouseMove]);
+    this.element.removeEventListener('mousedown', this[$onMouseDown]);
+    this.element.removeEventListener('wheel', this[$onWheel]);
+    this.element.removeEventListener('keydown', this[$onKeyDown]);
+    this.element.removeEventListener('touchstart', this[$onTouchStart]);
+    this.element.removeEventListener('touchmove', this[$onTouchMove]);
+
+    self.removeEventListener('mouseup', this[$onMouseUp]);
+    self.removeEventListener('touchend', this[$onTouchEnd]);
+  }
+
+  /**
+   * The constraints that are currently configured for the controls instance.
+   */
+  get constraints() {
+    return this[$constraints];
+  }
+
+  /**
+   * A copy of the spherical that represents the current camera position
+   * relative to the configured scene origin.
+   */
+  get cameraSpherical() {
+    return this[$spherical].clone();
```

Comment:
I think `getSpherical(target)` is more obvious of what's happening, rather than an unexpected allocation -- agreed that this shouldn't be modified by the consumer, my concern is more about the unexpected allocation (but again, only used in tests, so I don't feel strongly)

Reply:
Done